In [94]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [95]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                           1.08388                     7.60278   
ethereum                          0.22392                    10.38134   
tether                           -0.21173                     0.04935   
ripple                           -0.37819                    -0.60926   
bitcoin-cash                      2.90585                    17.09717   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                           6.57509                      7.67258   
ethereum                          4.80849                      0.13169   
tether                            0.00640                     -0.04237   
ripple                            2.24984                      0.23455   
bitcoin-cash                     14.75334                     15.74903   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                          -3.25185                      83.51840   
ethereum                        -12.88890                     186.77418   
tether                            0.28037                      -0.00542   
ripple                          -17.55245                      39.53888   
bitcoin-cash                    -13.71793                      21.66042   

              price_change_percentage_1y  
coin_id                                   
bitcoin                         37.51761  
ethereum                       101.96023  
tether                           0.01954  
ripple                         -16.60193  
bitcoin-cash                    14.49384

In [96]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [97]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [98]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [99]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_scaled = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y']])

# Diplay the first three rows of the scaled data
market_scaled[0:5]


array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [100]:
# Create a DataFrame with the scaled data
df_market_scaled = pd.DataFrame(market_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'])

df_market_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_scaled.set_index('coin_id',inplace=True)
# Display sample data
df_market_scaled.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

In [101]:
# Plot your data to see what's in your DataFrame
df_market_scaled.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

In [102]:
df_market_scaled.hvplot.scatter(x = "price_change_percentage_24h", 
                                 y = "price_change_percentage_7d",
                                 by = "coin_id"
                            )

:NdOverlay   [coin_id]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

### Find the Best Value for k Using the Original Data.

In [103]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [104]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
for i in k: # Inside the loop:
    k_model = KMeans(n_clusters=i, random_state=1) # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model.fit(df_market_scaled) # 2. Fit the model to the data using `df_market_data_scaled`
    inertia.append(k_model.inertia_) # 3. Append the model.inertia_ to the inertia list


c:\Users\Ina\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [105]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_data = pd.DataFrame.from_dict(elbow_data)
df_elbow_data

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.405923
5   6   52.933559
6   7   47.983124
7   8   37.288187
8   9   33.061685
9  10   28.779752

In [143]:
# Plot a line chart with all the inertia values computed with 
elbow_plot = df_elbow_data.hvplot.line(x = "k", y = "inertia",
    width=600,
    height=400,
    rot=90,
    title="Elbow Curve", 
    xticks=k
)
elbow_plot

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best K value is 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [107]:
# Initialise the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [108]:
# Fit the K-Means model using the scaled data
model.fit(df_market_scaled)

KMeans(n_clusters=4, random_state=0)

In [109]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_4 = model.predict(df_market_scaled)

# Print the resulting array of cluster values.
k_4

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [110]:
# Create a copy of the DataFrame
df_market_predict = df_market_scaled.copy()

In [111]:
# Add a new column to the DataFrame with the predicted clusters
df_market_predict["hover_cols"] = k_4

# Display sample data
df_market_predict.head(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  hover_cols  
coin_id                                               
bitcoin                        -0.251637           0  
ethereum                       -0.199352           0  
tether                         -0.282061           3  
ripple                         -0.295546           3  
bitcoin-cash                   -0.270317           0

In [149]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
market_plot = df_market_predict.hvplot.scatter(x = "price_change_percentage_24h", 
                                 y = "price_change_percentage_7d",
                                 by = "hover_cols"
                            )
market_plot

:NdOverlay   [hover_cols]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)

---

### Optimise Clusters with Principal Component Analysis.

In [113]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [114]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_pca = pca.fit_transform(df_market_scaled)
# View the first five rows of the DataFrame. 
market_pca[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [115]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** array([0.3719856 , 0.34700813, 0.17603793])

In [116]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data
df_market_pca = pd.DataFrame(market_pca,columns=["PC1","PC2","PC3"])

# Copy the crypto names from the original data
df_market_pca["coin_id"] = df_market_scaled.index

# Set the coinid column as index
df_market_pca.set_index("coin_id",inplace=True)

# Display sample data
df_market_pca.head(5)

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [117]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1,11))

In [118]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
for i in k_pca: # Inside the loop:
    k_model_pca = KMeans(n_clusters=i, random_state=1) # 1. Create a KMeans model using the loop counter for the n_clusters
    k_model_pca.fit(df_market_pca) # 2. Fit the model to the data using `df_market_data_scaled`
    inertia_pca.append(k_model_pca.inertia_) # 3. Append the model.inertia_ to the inertia list

c:\Users\Ina\anaconda3\envs\PythonData\lib\site-packages\sklearn\cluster\_kmeans.py:1037: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  "KMeans is known to have a memory leak on Windows "


In [123]:
# Create a dictionary with the data to plot the Elbow curve
elbow_pca_data = {"k_pca": k_pca, "inertia_pca": inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca_data = pd.DataFrame.from_dict(elbow_pca_data)
df_elbow_pca_data

k_pca  inertia_pca
0      1   256.874086
1      2   165.901994
2      3    93.774626
3      4    49.665497
4      5    38.352251
5      6    27.618972
6      7    21.134056
7      8    17.437664
8      9    13.742792
9     10    10.484890

In [142]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
# Plot a line chart with all the inertia values computed with 
elbow_pca_plot = df_elbow_pca_data.hvplot.line(x = 'k_pca', y = 'inertia_pca',
    width=600,
    height=400,
    rot=90,
    title="Elbow PCA Curve", 
    xticks=k
)
elbow_pca_plot

:Curve   [k_pca]   (inertia_pca)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for "k" is 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Same with the original data

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [127]:
# Initialise the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4,random_state=4) 

In [128]:
# Fit the K-Means model using the PCA data
model_pca.fit(df_market_pca)

KMeans(n_clusters=4, random_state=4)

In [130]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_pca_4 = model_pca.predict(df_market_pca)
# Print the resulting array of cluster values.
k_pca_4

array([0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 3, 0, 1, 1, 2, 1, 1, 1, 1])

In [146]:
# Create a copy of the DataFrame with the PCA data
market_pca_predict_df = df_market_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
market_pca_predict_df['hover_cols'] = k_pca_4
k_pca_4
# Display sample data
market_pca_predict_df.head(5)

PC1       PC2       PC3  hover_cols
coin_id                                               
bitcoin      -0.600667  0.842760  0.461595           0
ethereum     -0.458261  0.458466  0.952877           0
tether       -0.433070 -0.168126 -0.641752           1
ripple       -0.471835 -0.222660 -0.479053           1
bitcoin-cash -1.157800  2.041209  1.859715           0

In [148]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2". 
# Colour the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
market_pca_plot = market_pca_predict_df.hvplot.scatter(x = "PC1", 
                                     y = "PC2",
                                     by = "hover_cols"
                                   )
market_pca_plot

:NdOverlay   [hover_cols]
   :Scatter   [PC1]   (PC2)

In [151]:
market_pca_plot = market_pca_predict_df.hvplot.scatter(x = "PC2", 
                                     y = "PC3",
                                     by = "hover_cols"
                                   )
market_pca_plot

:NdOverlay   [hover_cols]
   :Scatter   [PC2]   (PC3)

In [152]:
market_pca_plot = market_pca_predict_df.hvplot.scatter(x = "PC1", 
                                     y = "PC3",
                                     by = "hover_cols"
                                   )
market_pca_plot

:NdOverlay   [hover_cols]
   :Scatter   [PC1]   (PC3)

### Visualise and Compare the Results

In this section, you will visually analyse the cluster analysis results by contrasting the outcome with and without using the optimisation techniques.

In [144]:
# Composite plot to contrast the Elbow curves
elbow_plot + elbow_pca_plot

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k_pca]   (inertia_pca)

In [150]:
# Composite plot to contrast the clusters
market_plot + market_pca_plot

:Layout
   .NdOverlay.I  :NdOverlay   [hover_cols]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .NdOverlay.II :NdOverlay   [hover_cols]
      :Scatter   [PC1]   (PC2)

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 